de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SharedReferencePointMapper.java

```diff
@@ -0,0 +1,543 @@
+     * @throws IllegalStateException
+     *             if the ID is already in use or the reference point was
+     *             already added
+     */
+    public synchronized void addReferencePoint(String id,
+        IReferencePoint referencePoint, boolean isPartially) {
+        boolean upgrade = false;
+
+        if (id == null)
+            throw new NullPointerException("ID is null");
+
+        if (referencePoint == null)
+            throw new NullPointerException("referencePoint is null");
+
+        String currentReferencePointID = referencePointToIDMapping
+            .get(referencePoint);
+        IReferencePoint currentReferencePoint = idToReferencePointMapping
+            .get(id);
+
+        if (currentReferencePointID != null
+            && !id.equals(currentReferencePointID)) {
+            throw new IllegalStateException("cannot assign ID " + id
+                + " to referencePoint " + referencePoint
+                + " because it is already registered with ID "
+                + currentReferencePointID);
+        }
+
+        if (currentReferencePoint != null
+            && !referencePoint.equals(currentReferencePoint)) {
+            throw new IllegalStateException("ID " + id + " for referencePoint "
+                + referencePoint + " is already used by referencePoint "
+                + currentReferencePoint);
+        }
+
+        if (isPartially
+            && partiallySharedReferencePoints.contains(referencePoint))
+            throw new IllegalStateException("referencePoint " + referencePoint
+                + " is already partially shared");
+
+        if (!isPartially
+            && completelySharedReferencePoints.contains(referencePoint))
+            throw new IllegalStateException("referencePoint " + referencePoint
+                + " is already completely shared");
+
+        if (isPartially
+            && completelySharedReferencePoints.contains(referencePoint))
+            throw new IllegalStateException(
+                "referencePoint "
+                    + referencePoint
+                    + " is already completely shared (cannot downgrade a completely shared referencePoint)");
+
+        if (!isPartially
+            && partiallySharedReferencePoints.contains(referencePoint)) {
+            partiallySharedReferencePoints.remove(referencePoint);
+            upgrade = true;
+        }
+
+        if (isPartially)
+            partiallySharedReferencePoints.add(referencePoint);
+        else
+            completelySharedReferencePoints.add(referencePoint);
+
+        assert Collections.disjoint(completelySharedReferencePoints,
+            partiallySharedReferencePoints);
+
+        if (upgrade) {
+            // release resources
+            partiallySharedResourceMapping.put(referencePoint, null);
+
+            LOG.debug("upgraded partially shared referencePoint "
+                + referencePoint + " with ID " + id
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Define a constant instead of duplicating this literal " with ID " 3 times. ([squid:S1192](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1192))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=257&issues=AWbeHvN3SgN4uCclRsGv&open=AWbeHvN3SgN4uCclRsGv)

[ISSUE_KEY]: AWbeHvN3SgN4uCclRsGv